### PR TITLE
Add minetest.register_on_punchplayer

### PR DESCRIPTION
**Pull request by TeTpaAka**
Rebase of https://github.com/minetest/minetest/pull/1307
Additionally adds a return value.
Return true to disable the damage handling.

### CODE REVIEW

File: doc/lua_api.txt [hunk 1]
```diff
@@ -2278,8 +2278,16 @@ These functions return the leftover itemstack.
 
 * `minetest.forceload_free_block(pos)`
     * stops forceloading the position `pos`
-
 Please note that forceloaded areas are saved when the server restarts.
+* `minetest.register_on_punchplayer(player, hitter, time_from_last_punch, tool_capabilities, dir)`
```
Comment (est31) [inline doc/lua_api.txt]: I'm missing a `func(` here.

**Comment by est31** [comment]
When you `return true` inside the function, then it still seems for the local player that the punched loses HP (turn debug output on with pressing f5 to get HP of other SAOs). Also, the log will show it up as actual damage, like:
```
ACTION[ServerThread]: Player a punched by player b, damage 1 HP
```

File: doc/lua_api.txt (hunk 1 above)
Comment (TeTpaAka) [inline doc/lua_api.txt]: Fixed. It was in the original pull request like this and I didn't see the mistake.

**Comment by TeTpaAka** [comment]
How can I force the send of an HP packet? The problem is, if I'm correct, that the HP packet is only send for an HP change (reasonable) but the client predicts the damage and applies it. Maybe the client should ask for a new HP packet after it predicts the damage?
BTW, the same happens for disabled damage on current master, so the damage bug is not completely my fault.

**Comment by est31** [comment]
Ah I see. Ok, then it isn't required.  :+1: if it gets squashed.

**Comment by TeTpaAka** [comment]
Squashed.

**Comment by TeTpaAka** [comment]
Changed the return from RUN_CALLBACKS_MODE_LAST to RUN_CALLBACKS_MODE_OR so any mods can log punch actions without re-enabling the damage. (Same commit)

**Comment by TeTpaAka** [comment]
I fixed the damage prediction and server log anyway. Should I squash again?

File: doc/lua_api.txt [hunk 2]
```diff
@@ -2278,8 +2278,16 @@ These functions return the leftover itemstack.
 
 * `minetest.forceload_free_block(pos)`
     * stops forceloading the position `pos`
-
 Please note that forceloaded areas are saved when the server restarts.
+* `minetest.register_on_punchplayer(func(player, hitter, time_from_last_punch, tool_capabilities, dir))`
```
Comment (rubenwardy) [inline doc/lua_api.txt]: This is in the wrong place, you should add it where all the other register_ons are.

Comment (TeTpaAka) [inline doc/lua_api.txt]: Sorry, fixed.

File: src/content_sao.cpp [hunk 3]
```diff
@@ -1037,11 +1037,33 @@ int PlayerSAO::punch(v3f dir,
...
+	PlayerSAO *playersao = m_player->getPlayerSAO();
+
+	bool damage_handled = m_env->getScriptIface()->on_punchplayer(playersao, puncher,
+				time_from_last_punch, toolcap, dir);
+
+	if (!damage_handled) {
+		setHP(getHP() - hitparams.hp);
+	}
+	else { // override client prediction
+		if(puncher->getType() == ACTIVEOBJECT_TYPE_PLAYER){
```
Comment (rubenwardy) [inline src/content_sao.cpp]: Should be

    if (cond) {

Notice spaces.

File: src/content_sao.cpp [hunk 4]
```diff
@@ -1037,11 +1037,33 @@ int PlayerSAO::punch(v3f dir,
...
-			<<punchername<<", damage "<<hitparams.hp
-			<<" HP"<<std::endl;
+	PlayerSAO *playersao = m_player->getPlayerSAO();
+
+	bool damage_handled = m_env->getScriptIface()->on_punchplayer(playersao, puncher,
+				time_from_last_punch, toolcap, dir);
+
+	if (!damage_handled) {
+		setHP(getHP() - hitparams.hp);
+	}
```
Comment (rubenwardy) [inline src/content_sao.cpp]: Should be 

   } else {

File: src/content_sao.cpp [hunk 5]
```diff
@@ -1037,11 +1037,33 @@ int PlayerSAO::punch(v3f dir,
...
 
-	setHP(getHP() - hitparams.hp);
+	
+	actionstream<<"Player "<<m_player->getName()<<" punched by "
+			<<punchername;
+	if (!damage_handled) {
+		actionstream<<", damage "<<hitparams.hp<<" HP"; //
+	}
+	else {
```
Comment (rubenwardy) [inline src/content_sao.cpp]: And here.

File: src/script/cpp_api/s_player.cpp [hunk 6]
```diff
@@ -45,6 +47,27 @@ void ScriptApiPlayer::on_dieplayer(ServerActiveObject *player)
...
+	lua_getfield(L, -1, "registered_on_punchplayer");
+	// Call callbacks
+	objectrefGetOrCreate(L, player);
+	objectrefGetOrCreate(L, hitter);
+	lua_pushnumber(L, time_from_last_punch);
+	push_tool_capabilities(L, *toolcap);
+	push_v3f(L, dir);
+	script_run_callbacks(L, 5, RUN_CALLBACKS_MODE_OR);
+	bool damage_handled = lua_toboolean(L, -1);
+	return damage_handled;
```
Comment (rubenwardy) [inline src/script/cpp_api/s_player.cpp]: Seems a bit weird to create a variable here.

File: src/script/cpp_api/s_player.h [hunk 7]
```diff
@@ -38,6 +40,7 @@ class ScriptApiPlayer
 	void on_joinplayer(ServerActiveObject *player);
 	void on_leaveplayer(ServerActiveObject *player);
 	void on_cheat(ServerActiveObject *player, const std::string &cheat_type);
+	bool on_punchplayer(ServerActiveObject *player, ServerActiveObject *hitter,float time_from_last_punch, const ToolCapabilities *toolcap, v3f dir);
```
Comment (rubenwardy) [inline src/script/cpp_api/s_player.h]: Missing a space:

     hitter,float


Don't worry, code style issues aren't that big.

File: src/content_sao.cpp (hunk 3 above)
Comment (TeTpaAka) [inline src/content_sao.cpp]: Sure. That's the way I code normally, but I simply copied the code from above. Fixed.

File: src/content_sao.cpp (hunk 4 above)
Comment (TeTpaAka) [inline src/content_sao.cpp]: Fixed.

File: src/content_sao.cpp (hunk 5 above)
Comment (TeTpaAka) [inline src/content_sao.cpp]: Fixed

File: src/script/cpp_api/s_player.cpp (hunk 6 above)
Comment (TeTpaAka) [inline src/script/cpp_api/s_player.cpp]: It was done that way in the function right below. But I changed it anyway.

File: src/script/cpp_api/s_player.h (hunk 7 above)
Comment (TeTpaAka) [inline src/script/cpp_api/s_player.h]: Fixed.

**Comment by rubenwardy** [comment]
As a general rule, don't look at surrounding code for a style guide.

People usually correct code around their patches, but that doesn't matter much. (Not usually a condition of merging)

File: src/script/cpp_api/s_player.cpp (hunk 6 above)
Comment (est31) [inline src/script/cpp_api/s_player.cpp]: Both would have been ok.

**Comment by est31** [comment]
Style guide [here](http://dev.minetest.net/Code_style_guidelines).

File: src/content_sao.cpp [hunk 8]
```diff
@@ -1034,14 +1034,34 @@ int PlayerSAO::punch(v3f dir,
...
+		if (puncher->getType() == ACTIVEOBJECT_TYPE_PLAYER) {
+			std::string str = gob_cmd_punched(0, getHP());
+			// create message and add to list
+			ActiveObjectMessage aom(getId(), true, str);
+			m_messages_out.push(aom);
+		}
+	}
+
+	
+	actionstream<<"Player "<<m_player->getName()<<" punched by "
```
Comment (est31) [inline src/content_sao.cpp]: There should be spaces before and after each `<<`.

File: src/script/cpp_api/s_player.h [hunk 9]
```diff
@@ -38,6 +40,7 @@ class ScriptApiPlayer
 	void on_joinplayer(ServerActiveObject *player);
 	void on_leaveplayer(ServerActiveObject *player);
 	void on_cheat(ServerActiveObject *player, const std::string &cheat_type);
+	bool on_punchplayer(ServerActiveObject *player, ServerActiveObject *hitter, float time_from_last_punch, const ToolCapabilities *toolcap, v3f dir);
```
Comment (est31) [inline src/script/cpp_api/s_player.h]: That line is too long. Every line should be shorter than 70 chars, but its ok to have lines up to 89 chars.

Comment (TeTpaAka) [inline src/script/cpp_api/s_player.h]: Fixed.

File: src/content_sao.cpp (hunk 8 above)
Comment (TeTpaAka) [inline src/content_sao.cpp]: Fixed.

**Comment by est31** [comment]
Very nice. I've already :+1: ed it.

**Comment by TeTpaAka** [comment]
Maybe I should also pass the inflicted damage? I just realized it was missing.

File: src/content_sao.cpp [hunk 10]
```diff
@@ -1034,14 +1034,34 @@ int PlayerSAO::punch(v3f dir,
...
+
+	
+	actionstream << "Player " << m_player->getName() << " punched by "
+			<< punchername;
+	if (!damage_handled) {
+		actionstream << ", damage " << hitparams.hp << " HP";
+	} else {
+		actionstream << ", damage handled by lua";
+	}
+	actionstream<<std::endl;
```
Comment (est31) [inline src/content_sao.cpp]: spaces :)

Comment (TeTpaAka) [inline src/content_sao.cpp]: Missed it. Now fixed.

**Comment by Zeno-** [comment]
Always squash the commits if the "function"/intent of the commit remains the same. E.g. something that fixes a typo or style issue in your PR should always be squashed

**Comment by TeTpaAka** [comment]
Also pass the damage value to the on_punchplayer function and squashed everything in one commit.

**Comment by TeTpaAka** [comment]
Fixed wrong comment, stating on_dieplayers and fixing the table name to core.registered_on_punchplayers (the s was missing).

**Comment by Novatux** [comment]
:+1:

File: src/script/cpp_api/s_player.h [hunk 11]
```diff
@@ -38,6 +40,9 @@ class ScriptApiPlayer
 	void on_joinplayer(ServerActiveObject *player);
 	void on_leaveplayer(ServerActiveObject *player);
 	void on_cheat(ServerActiveObject *player, const std::string &cheat_type);
+	bool on_punchplayer(ServerActiveObject *player,
```
Comment (est31) [inline src/script/cpp_api/s_player.h]: remove trailing whitespace please

File: src/content_sao.cpp [hunk 12]
```diff
@@ -1034,14 +1034,35 @@ int PlayerSAO::punch(v3f dir,
...
+	} else { // override client prediction
+		if (puncher->getType() == ACTIVEOBJECT_TYPE_PLAYER) {
+			std::string str = gob_cmd_punched(0, getHP());
+			// create message and add to list
+			ActiveObjectMessage aom(getId(), true, str);
+			m_messages_out.push(aom);
+		}
+	}
+
+
```
Comment (est31) [inline src/content_sao.cpp]: trailing whitespace

**Comment by est31** [comment]
removed them. c5b4e541749c50805519ce040d98a0a8e5e0ec03

**Pull request closed without merging**